File: web/src/lib/repair.ts
/**
 * 修復用フレーム（REPAIR）: 複数のチャンクを XOR で重ねたもの（Python 版 src/qrtransfer/repair.py と同じ決まり）。
 * 受け取った DATA と REPAIR の枚数が全チャンク数を少し超えた時点で、欠けたチャンクを計算で求められる。
 *
 * どのチャンクを重ねるかは session_id と修復用フレームの番号から、整数演算だけで決まる（Python 版と必ず一致する）。
 */

export const DEGREE_MAX = 1024;
export const TOTAL_LIMIT = 2 ** 21; // 番号の計算（next() * n）を 2^53 未満に収めるため
const MEMORY_LIMIT = 64 * 1024 * 1024; // まだ解けていない式を保持する上限（バイト）

export function degree(total: number): number {
  return Math.max(1, Math.min(total, DEGREE_MAX, Math.floor((total + 1) / 2)));
}

class Mulberry32 {
  private a: number;

  constructor(seed: number) {
    this.a = seed >>> 0;
  }

  next(): number {
    this.a = (this.a + 0x6d2b79f5) >>> 0;
    let t = this.a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return (t ^ (t >>> 14)) >>> 0;
  }

  below(n: number): number {
    return Math.floor((this.next() * n) / 4294967296);
  }
}

/** index 番目の修復用フレームが重ねるチャンクの番号（昇順）。 */
export function repairIndices(sessionId: number, index: number, total: number): number[] {
  if (!(total > 0 && total < TOTAL_LIMIT)) throw new RangeError(`total out of range: ${total}`);
  const rng = new Mulberry32((sessionId ^ Math.imul(index + 1, 0x9e3779b1)) >>> 0);
  const d = degree(total);
  const chosen = new Set<number>();
  // Floyd の方法: total 個から d 個を重複なく選ぶ
  for (let j = total - d; j < total; j++) {
    const t = rng.below(j + 1);
    chosen.add(chosen.has(t) ? j : t);
  }
  return [...chosen].sort((x, y) => x - y);
}

interface Row {
  bits: Uint32Array; // 未知数（まだ受け取っていないチャンク）のビット列
  data: Uint8Array;
}

function xorInto(dst: Uint32Array, src: Uint32Array): void {
  for (let i = 0; i < dst.length; i++) dst[i] ^= src[i];
}

/** dst ^= src（src の長さまで）。両方が 4 バイト境界にあれば 4 バイトずつ計算する（1 バイトずつの約 4 倍速い）。 */
function xorBytes(dst: Uint8Array, src: Uint8Array): void {
  const n = Math.min(dst.length, src.length);
  let i = 0;
  if (dst.byteOffset % 4 === 0 && src.byteOffset % 4 === 0) {
    const words = n >>> 2;
    const d = new Uint32Array(dst.buffer, dst.byteOffset, words);
    const s = new Uint32Array(src.buffer, src.byteOffset, words);
    for (let k = 0; k < words; k++) d[k] ^= s[k];
    i = words * 4;
  }
  for (; i < n; i++) dst[i] ^= src[i];
}

function lowestBit(bits: Uint32Array): number {
  for (let w = 0; w < bits.length; w++) {
    const v = bits[w];
    if (v) return w * 32 + (31 - Math.clz32(v & -v));
  }
  return -1;
}

function isSingle(bits: Uint32Array): boolean {
  let seen = false;
  for (let w = 0; w < bits.length; w++) {
    const v = bits[w];
    if (!v) continue;
    if (seen || v & (v - 1)) return false;
    seen = true;
  }
  return seen;
}

/**
 * 修復用フレームの式を集め、解けたチャンクを返す（Python 版 repair.Decoder と同じ手順）。
 * 式は既約行階段形（各行の先頭の未知数は他の行に現れない）で持ち、未知数が 1 つになった行が「解けた」チャンク。
 */
export class RepairDecoder {
  private rows = new Map<number, Row>();
  private readonly words: number;
  readonly maxRows: number;
  dropped = 0;

  constructor(
    readonly total: number,
    readonly length: number,
    memoryLimit = MEMORY_LIMIT,
  ) {
    this.words = Math.ceil(total / 32);
    this.maxRows = Math.max(1, Math.floor(memoryLimit / (this.words * 4 + length + 64)));
  }

  get pending(): number {
    return this.rows.size;
  }

  private padded(data: Uint8Array): Uint8Array {
    const out = new Uint8Array(this.length);
    out.set(data.subarray(0, this.length));
    return out;
  }

  /** 修復用フレーム 1 枚分の式を加える。known(i) は受信済みのチャンク。解けたチャンクを返す。 */
  addRepair(indices: number[], payload: Uint8Array, known: (i: number) => Uint8Array | undefined): [number, Uint8Array][] {
    const bits = new Uint32Array(this.words);
    const data = this.padded(payload);
    for (const i of indices) {
      const c = known(i);
      if (c) {
        xorBytes(data, c);
      } else {
        bits[i >>> 5] |= 1 << (i & 31);
      }
    }
    return this.insert(bits, data, -1);
  }

  /** DATA で受け取ったチャンクを式に反映する。これによって解けた（ほかの）チャンクを返す。 */
  addKnown(seq: number, chunk: Uint8Array): [number, Uint8Array][] {
    const w = seq >>> 5;
    const m = 1 << (seq & 31);
    let used = false;
    for (const r of this.rows.values()) {
      if (r.bits[w] & m) {
        used = true;
        break;
      }
    }
    if (!used) return [];
    const bits = new Uint32Array(this.words);
    bits[w] = m;
    return this.insert(bits, this.padded(chunk), seq);
  }

  private insert(bits: Uint32Array, data: Uint8Array, skip: number): [number, Uint8Array][] {
    // 既にある行の先頭の未知数を消す（各行の先頭は他の行に現れないので、元の式の位置だけ見ればよい）
    const orig = bits.slice();
    for (let w = 0; w < orig.length; w++) {
      let v = orig[w];
      while (v) {
        const low = v & -v;
        v ^= low;
        const row = this.rows.get(w * 32 + (31 - Math.clz32(low)));
        if (row) {
          xorInto(bits, row.bits);
          xorBytes(data, row.data);
        }
      }
    }
    const pivot = lowestBit(bits);
    if (pivot < 0) return []; // ほかの式から導ける
    if (this.rows.size >= this.maxRows && !isSingle(bits)) {
      this.dropped++;
      return [];
    }
    const w = pivot >>> 5;
    const m = 1 << (pivot & 31);
    const touched = [pivot];
    for (const [q, r] of this.rows) {
      if (r.bits[w] & m) {
        xorInto(r.bits, bits);
        xorBytes(r.data, data);
        touched.push(q);
      }
    }
    this.rows.set(pivot, { bits, data });
    const solved: [number, Uint8Array][] = [];
    for (const q of touched) {
      const r = this.rows.get(q)!;
      if (isSingle(r.bits)) {
        this.rows.delete(q);
        if (q !== skip) solved.push([q, r.data]);
      }
    }
    return solved;
  }
}

File: src/qrtransfer/repair.py
"""修復用フレーム（REPAIR）: 複数のチャンクを XOR で重ねたもの。欠けたチャンクを受信側で計算して復元する。

送信側は DATA を 1 周流したあと、修復用フレームを流す（以後はその繰り返し）。受信側は、どのフレームを
取りこぼしても、受け取った DATA と REPAIR の枚数が全チャンク数を少し超えた時点で、残りを計算で求められる
（欠落番号を送信側に伝えて再送してもらう必要がない）。

どのチャンクを重ねるかは、session_id と修復用フレームの番号から決まる（Web 版 web/src/lib/repair.ts と同じ手順。
浮動小数点を使わない整数演算だけで決めるので、Python と JavaScript で必ず一致する）。
"""

from __future__ import annotations

import math
from typing import Callable, Iterable

import numpy as np

_M = 0xFFFFFFFF

DEGREE_MAX = 1024  # 1 枚の修復用フレームに重ねるチャンク数の上限
TOTAL_LIMIT = 1 << 21  # これより多いチャンク数の転送では修復用フレームを使わない（番号の計算を整数の範囲に収める）
RATIO_DEFAULT = 50  # 修復用フレームの枚数（全チャンク数に対する %）
RATIO_MAX = 200
MEMORY_LIMIT = 128 * 1024 * 1024  # 受信側で、まだ解けていない式を保持する上限（バイト）


def degree(total: int) -> int:
    """重ねるチャンク数。少ないときは半分（どの欠落にも効く）、多いときは DEGREE_MAX。"""
    return max(1, min(total, DEGREE_MAX, (total + 1) // 2))


def repair_count(total: int, ratio: int) -> int:
    """送信側で作る修復用フレームの枚数。ratio=0（従来方式）や大きすぎる転送では 0。"""
    if ratio <= 0 or total <= 0 or total >= TOTAL_LIMIT:
        return 0
    return max(4, math.ceil(total * ratio / 100))


class _Mulberry32:
    def __init__(self, seed: int):
        self.a = seed & _M

    def next(self) -> int:
        self.a = (self.a + 0x6D2B79F5) & _M
        t = self.a
        t = ((t ^ (t >> 15)) * (t | 1)) & _M
        t ^= (t + ((t ^ (t >> 7)) * (t | 61))) & _M
        return (t ^ (t >> 14)) & _M

    def below(self, n: int) -> int:
        return (self.next() * n) >> 32


def indices(session_id: int, index: int, total: int) -> list[int]:
    """index 番目の修復用フレームが重ねるチャンクの番号（昇順）。"""
    if not 0 < total < TOTAL_LIMIT:
        raise ValueError(f"total out of range: {total}")
    rng = _Mulberry32(session_id ^ (((index + 1) * 0x9E3779B1) & _M))
    d = degree(total)
    chosen: set[int] = set()
    # Floyd の方法: total 個から d 個を重複なく選ぶ
    for j in range(total - d, total):
        t = rng.below(j + 1)
        chosen.add(j if t in chosen else t)
    return sorted(chosen)


def encode(chunks: list[bytes], chunk_size: int, session_id: int, count: int) -> list[bytes]:
    """修復用フレームの payload（長さ chunk_size。短い最終チャンクは 0 で埋めて重ねる）を count 枚作る。"""
    total = len(chunks)
    if count <= 0 or total == 0:
        return []
    arr = np.zeros((total, chunk_size), dtype=np.uint8)
    for i, c in enumerate(chunks):
        arr[i, :len(c)] = np.frombuffer(c, dtype=np.uint8)
    return [np.bitwise_xor.reduce(arr[indices(session_id, r, total)], axis=0).tobytes() for r in range(count)]


class Decoder:
    """修復用フレームの式（どのチャンクの XOR か）を集め、解けたチャンクを返す。

    式は「まだ受け取っていないチャンク」だけを未知数として、既約行階段形（各行の先頭の未知数は他の行に現れない）で持つ。
    未知数が 1 つだけになった行は、そのチャンクが解けたことを表す。
    ビット列と payload は Python の int で持つ（XOR が速い）。
    """

    def __init__(self, total: int, length: int, memory_limit: int = MEMORY_LIMIT):
        self.total = total
        self.length = length
        self.rows: dict[int, tuple[int, int]] = {}  # 先頭の未知数 → (未知数のビット列, payload)
        self.max_rows = max(1, memory_limit // (total // 8 + length + 64))
        self.dropped = 0  # 保持の上限で捨てた式の数（診断用）

    def _pad(self, data: bytes) -> int:
        return int.from_bytes(bytes(data).ljust(self.length, b"\0"), "big")

    def add_repair(self, idx: Iterable[int], payload: bytes, known: Callable[[int], bytes | None]) -> list[tuple[int, bytes]]:
        """修復用フレーム 1 枚分の式を加える。known(i) は受信済みのチャンク（なければ None）。解けたチャンクを返す。"""
        bits = 0
        data = self._pad(payload)
        for i in idx:
            c = known(i)
            if c is None:
                bits |= 1 << i
            else:
                data ^= self._pad(c)
        return self._insert(bits, data, None)

    def add_known(self, seq: int, chunk: bytes) -> list[tuple[int, bytes]]:
        """DATA で受け取ったチャンクを式に反映する。これによって解けた（ほかの）チャンクを返す。"""
        mask = 1 << seq
        if not any(b & mask for b, _ in self.rows.values()):
            return []
        return self._insert(mask, self._pad(chunk), seq)

    def _insert(self, bits: int, data: int, skip: int | None) -> list[tuple[int, bytes]]:
        x = bits
        while x:
            low = x & -x
            x ^= low
            row = self.rows.get(low.bit_length() - 1)
            if row is not None:
                bits ^= row[0]
                data ^= row[1]
        if not bits:
            return []  # ほかの式から導ける（新しい情報がない）
        if len(self.rows) >= self.max_rows and bits & (bits - 1):
            self.dropped += 1
            return []
        low = bits & -bits
        pivot = low.bit_length() - 1
        touched = [pivot]
        for q, (b, d) in self.rows.items():
            if b & low:
                self.rows[q] = (b ^ bits, d ^ data)
                touched.append(q)
        self.rows[pivot] = (bits, data)
        solved: list[tuple[int, bytes]] = []
        for q in touched:
            b, d = self.rows[q]
            if not b & (b - 1):
                del self.rows[q]
                if q != skip:
                    solved.append((q, d.to_bytes(self.length, "big")))
        return solved
